rules/prefer-class-fields.js

```diff
@@ -0,0 +1,182 @@
+*/
+const isThisAssignmentExpression = node => {
+	if (
+		node.type !== 'ExpressionStatement'
+		|| node.expression.type !== 'AssignmentExpression'
+	) {
+		return false;
+	}
+
+	const lhs = node.expression.left;
+
+	if (!lhs.object || lhs.object.type !== 'ThisExpression') {
+		return false;
+	}
+
+	return true;
+};
+
+/**
+@param {import('eslint').Rule.Node} node
+@param {import('eslint').Rule.RuleContext['sourceCode']} sourceCode
+@param {import('eslint').Rule.RuleFixer} fixer
+*/
+const removeFieldAssignment = (node, sourceCode, fixer) => {
+	const {line} = node.loc.start;
+	const nodeText = sourceCode.getText(node);
+	const lineText = sourceCode.lines[line - 1];
+	const isOnlyNodeOnLine = lineText.trim() === nodeText;
+
+	return isOnlyNodeOnLine
+		? fixer.removeRange([
+			sourceCode.getIndexFromLoc({line, column: 0}),
+			sourceCode.getIndexFromLoc({line: line + 1, column: 0}),
+		])
+		: fixer.remove(node);
+};
+
+/**
+@param {string} propertyName
+@param {import('estree').ClassBody} classBody
+*/
+const findClassFieldNamed = (propertyName, classBody) => {
+	for (const classBodyChild of classBody.body) {
+		if (
+			classBodyChild.type === 'PropertyDefinition'
+			&& classBodyChild.key.type === 'Identifier'
+			&& classBodyChild.key.name === propertyName
+		) {
+			return classBodyChild;
+		}
+	}
+};
+
+/**
+@param {string} propertyName
+@param {string} propertyValue
+@param {import('estree').ClassBody} classBody
+@param {import('estree').MethodDefinition} constructor
+@param {import('eslint').Rule.RuleContext['sourceCode']} sourceCode
+@param {import('eslint').Rule.RuleFixer} fixer
+*/
+const addOrReplaceClassFieldDeclaration = (
+	propertyName,
+	propertyValue,
+	classBody,
+	constructor,
+	sourceCode,
+	fixer,
+) => {
+	const alreadyExistingDeclaration = findClassFieldNamed(
+		propertyName,
+		classBody,
+	);
+
+	if (alreadyExistingDeclaration) {
+		return fixer.replaceText(
+			alreadyExistingDeclaration,
+			`${propertyName} = ${propertyValue};`,
+		);
+	}
+
+	const classBodyStartRange = [classBody.range[0], classBody.range[0] + 1];
+	const indent = getIndentString(constructor, sourceCode);
+	return fixer.insertTextAfterRange(
+		classBodyStartRange,
+		`\n${indent}${propertyName} = ${propertyValue};`,
+	);
+};
+
+/**
+@type {import('eslint').Rule.RuleModule['create']}
+*/
+const create = context => {
+	const {sourceCode} = context;
+
+	return {
+		ClassBody(classBody) {
+			const constructor = classBody.body.find(x => x.kind === 'constructor');
+
+			if (!constructor || constructor.type !== 'MethodDefinition') {
+				return;
+			}
+
+			const constructorBody = constructor.value.body?.body;
+
+			if (!constructorBody) {
+				return;
+			}
+
+			const firstInvalidProperty = constructorBody.findIndex(
+				node => node.type !== 'ExpressionStatement',
+			);
```

Comment:
Please help me understand this, are you assuming assignments before non-`ExpressionStatement`s are reachable? And it's addressing https://github.com/sindresorhus/eslint-plugin-unicorn/pull/2512#issuecomment-2547309562 ?

Reply:
I'm not sure about this, but I don't have a better idea. 

Anyway, let's allow `EmptyStatement` too. since in no-semi style, it's common to put leading `;` at start.

Reply:
done 7863af915754a4ccf6fc5689ad365c5409eea3ed